pkg/operations/kubernetesupgrade/upgradecluster.go

```diff
@@ -250,16 +207,19 @@ func (uc *UpgradeCluster) getClusterNodeStatus(subscriptionID uuid.UUID, az armh
 	return nil
 }
 
-// getClusterNodeVersion returns a node's "orchestrator:version" via Kubernetes API or VM tag.
-func (uc *UpgradeCluster) getClusterNodeVersion(client armhelpers.KubernetesClient, name string, tags map[string]*string) string {
+// getNodeVersion returns a node's current Kubernetes version via Kubernetes API or VM tag.
+func (uc *UpgradeCluster) getNodeVersion(client armhelpers.KubernetesClient, name string, tags map[string]*string) string {
```

Comment:
Can we reverse the precedence order here? Shouldn't we prefer to check the k8s API first, and then fall back to the tag?

Reply:
This is the same behavior that was implemented in @mboersma's PR to stop requiring tags https://github.com/Azure/acs-engine/pull/4100 so I'll defer to him for choosing this approach over the other.

I think the idea was that we maintain existing behavior to limit risk and only use the alternative method if we can't find the tags. 

If we do decide we want to reverse the order, I would argue that's a behavior change and it doesn't belong in this refactor PR (which is already dangerous enough as is).

Reply:
I agree with Cécile's comment here. I think it's safer to leave it alone.

This way the existing behavior is still the default (use tags), and the k8s API is explicitly a fallback strategy. At the time, that was deemed to be less disruptive, but it has proved to be useful in failed upgrade cases: setting the tags is a safe way to "trick" the upgrade behavior to run again.